### PR TITLE
Refactor bubble

### PR DESCRIPTION
**Pull request by cathmer4**
Changes the way bubbles split: they now get a list of bubbles that will appear when the bubble splits. This makes it easier to change the amount of bubbles that come out of a split. 
Also overloaded the createBubble method of the bubble factory. One method doesn't have any parameters and will create a bubble at location 0,0; the other method takes location x and y as input and creates a bubble at this location.

### CODE REVIEW

File: src/main/java/nl/tudelft/model/bubble/AbstractBubble.java [hunk 1]
```diff
@@ -29,7 +31,7 @@
     private boolean frozen = false;
     private int tickCount = 0;
     private final ResourcesWrapper resources;
-    private BubbleFactory bubbleFactory = null;
+    private List<AbstractBubble> next;
```
Comment (Pathemeous) [inline src/main/java/nl/tudelft/model/bubble/AbstractBubble.java]: This list's name is rather vague, nextBubbles seems better fit.

< naming

Comment (cathmer4) [inline src/main/java/nl/tudelft/model/bubble/AbstractBubble.java]: Changed

< naming

**Comment by Pathemeous** [comment]
How about creating an abstract method in AbstractBubble that must be defined in all bubbles to return the correct nextBubble list? Right now you call setNext() in the sub-constructor, which is not safe: an extension may forget about this and you end up with a null field. Creating an abstract method also enforces adherence to the contract (i.e. AbstractBubble uses the abstract method to define the nextBubbles list, instead of sub-bubbles setting the field for it.)

In the AbstractBubble constructor, we can then easily do the following:
<code>this.nextBubbles = abstractNextBubbleListCreator();</code>

Every bubble must implement it, and it automatically is incorporated in the functionality of the AbstractBubble and its split() method.

< organisation

**Comment by cathmer4** [comment]
Good idea, will do!

< organisation

**Comment by cathmer4** [comment]
I changed it. Also added an abstract method for the maxverticalspeed, because it is also essential that this is set.

**Comment by Pathemeous** [comment]
Great! I checkout'ed the branch to really have a good overview of what's going on. I added some hyperlinks to the JavaDoc and made it more specific for the <code>initMaxVerticalSpeed</code>. Namely: it should return a <b>positive</b> float value.
Of course, if you disagree, feel free to revert or amend my commits. I only did it because it was only a few keys away, and it felt silly to come back here and ask you to do it, which would take more time.

My final last point is the following, which I only notice now, but is definitely part of refactoring the Bubble classes:
The concrete bubbleTest classes do not extend the Abstract one. This was apparently not really relevant, but now with the abstract methods, there is a specified contract that <code>AbstractBubble</code> can test. 

Example: does the <code>initMaxVerticalSpeed</code> return a float value that is greater than 0?

Could you add this test case to the abstract Bubble and make the other test classes extend the AbstractBubbleTest class?

< testing, review

**Comment by cathmer4** [comment]
I'm not sure how to do this. I'm also not sure that it's worth the effort, because it's a minor improvement in testing, which doesn't have a high priority.

**Comment by Pathemeous** [comment]
It would add a little to the robustness of the unit tests. But I agree that this is not worth the time right now, especially since the issues earlier today #(182).

Merging.

**Pull request merged**